subsys/bluetooth/mesh/prov.c

```diff
@@ -971,20 +1107,101 @@ static void prov_input_complete(const u8_t *data)
 	notify_input_complete();
 }
 
-static void prov_confirm(const u8_t *data)
+#if defined(CONFIG_BT_MESH_PROVISIONER)
+static void send_prov_data(void)
 {
-	BT_DBG("Remote Confirm: %s", bt_hex(data, 16));
+	PROV_BUF(pdu, 34);
+	struct bt_mesh_subnet *sub;
+	u8_t session_key[16];
+	u8_t nonce[13];
+	u8_t data[25];
+	int err;
 
-	memcpy(link.conf, data, 16);
+	err = bt_mesh_session_key(link.dhkey, link.prov_salt, session_key);
+	if (err) {
+		BT_ERR("Unable to generate session key");
+		prov_send_fail_msg(PROV_ERR_UNEXP_ERR);
+		return;
+	}
 
-	notify_input_complete();
-	send_confirm();
+	BT_DBG("SessionKey: %s", bt_hex(session_key, 16));
+
+	err = bt_mesh_prov_nonce(link.dhkey, link.prov_salt, nonce);
+	if (err) {
+		BT_ERR("Unable to generate session nonce");
+		prov_send_fail_msg(PROV_ERR_UNEXP_ERR);
+		return;
+	}
+
+	BT_DBG("Nonce: %s", bt_hex(nonce, 13));
+
+	err = bt_mesh_dev_key(link.dhkey, link.prov_salt, link.node->dev_key);
+	if (err) {
+		BT_ERR("Unable to generate device key");
+		prov_send_fail_msg(PROV_ERR_UNEXP_ERR);
+		return;
+	}
+
+	BT_DBG("DevKey: %s", bt_hex(link.node->dev_key, 16));
+
+	sub = bt_mesh_subnet_get(link.node->net_idx);
+	if (sub == NULL) {
+		BT_ERR("No subnet with net_idx %u", link.node->net_idx);
+		prov_send_fail_msg(PROV_ERR_UNEXP_ERR);
+		return;
+	}
+
+	memcpy(data, sub->keys[sub->kr_flag].net, 16);
+	sys_put_be16(link.node->net_idx, &data[16]);
+	data[18] = bt_mesh_net_flags(sub);
+	sys_put_be32(bt_mesh.iv_index, &data[19]);
+	sys_put_be16(link.node->addr, &data[23]);
+
+	BT_DBG("net_idx %u, iv_index 0x%08x, addr 0x%04x",
+	       link.node->net_idx, bt_mesh.iv_index, link.node->addr);
+
+	prov_buf_init(&pdu, PROV_DATA);
+
+	err = bt_mesh_prov_encrypt(session_key, nonce, data,
+				   net_buf_simple_add(&pdu, 33));
+	if (err) {
+		BT_ERR("Unable to encrypt provisioning data");
+		prov_send_fail_msg(PROV_ERR_DECRYPT);
+		return;
+	}
+
+	if (prov_send(&pdu)) {
+		BT_ERR("Failed to send Provisioning Data");
+		return;
+	}
+
+	link.expect = PROV_COMPLETE;
 }
+#endif /* CONFIG_BT_MESH_PROVISIONER */
 
-static void prov_random(const u8_t *data)
+static void send_random(void)
 {
 	PROV_BUF(rnd, 16);
```

Comment:
This buffer is one byte too small to fit the random data plus packet type. Looks like an existing bug.

Reply:
Fixed, good catch :)

Reply:
This has been addressed in #19789, btw.